<launch>
    <arg name="record" default="false" doc="Set this flag to true to call record.launch"/>
    <arg name="bag_filename" default="recording.bag" doc="Name of the output ROS bag file in case record is set to true"/>
    <arg name="use_ned_frame" default="false" doc="If true, uses the NED (North-East-Down) frame conversion. If false, ENU (East-North-Up) will be used per default."/>

    <arg name="x" default="125" doc="X coordinate of the vehicle's initial position (in ENU)"/>
    <arg name="y" default="25" doc="Y coordinate of the vehicle's initial position (in ENU)"/>
    <arg name="z" default="-76" doc="Z coordinate of the vehicle's initial position (in ENU)"/>
    <arg name="yaw" default="-2.83" doc="Yaw angle of the vehicle's initial orientation"/>

    <arg name="teleop_on" default="false" doc="If true, the teleop node will be started"/>
    <arg name="joy_id" default="0" doc="Joystick ID"/>
    
    <arg name="namespace" default="rexrov2" doc="Namespace to spawn the vehicle"/>

    <arg name="sonar_name" default="blueview_p900"/>
    <arg name="sonar_image_topic" default="sonar_image"/>

    <include file="$(find uuv_gazebo_worlds)/launch/ocean_waves.launch"/>

    <include file="$(find uuv_simulation_wrapper)/launch/unpause_simulation.launch">
        <arg name="timeout" value="5"/>
    </include>

    <include file="$(find rexrov2_description)/launch/upload_rexrov2.launch">
        <arg name="x" value="$(arg x)"/>
        <arg name="y" value="$(arg y)"/>
        <arg name="z" value="$(arg z)"/>
        <arg name="yaw" value="$(arg yaw)"/>
        <arg name="use_ned_frame" value="$(arg use_ned_frame)"/>
        <arg name="sonar_name" value="$(arg sonar_name)"/>
        <arg name="gpu_ray" value="true"/>
        <arg name="maxDistance" value="15"/>
        
        <arg name="fidelity" value="200"/> 
        
        <arg name="raySkips" value="10"/>
        <arg name="sonar_image_topic" value="$(arg sonar_image_topic)"/>
        <arg name="sonar_image_raw_topic" value="sonar_image_raw"/>
        <arg name="plotScaler" value="1"/>
        <arg name="sensorGain" value="0.04"/>
        <arg name="ray_visual" value="false"/>
        <!-- <arg name="ray_visual" value="true"/> -->
        <arg name="writeLog" value="true"/>
        <arg name="writeFrameInterval" value="1"/>
    </include>

    <include file="$(find rexrov2_control)/launch/start_pid_controller.launch">
        <arg name="teleop_on" value="$(arg teleop_on)"/>
        <arg name="joy_id" value="$(arg joy_id)"/>
        <arg name="use_ned_frame" value="$(arg use_ned_frame)"/>
        <arg name="gui_on" value="false"/>
    </include>

    <!-- <node pkg="obstacle_avoidance" type="sonar_to_cloud.py" name="sonar_to_cloud" output="screen" /> -->
    <node pkg="obstacle_avoidance" type="sonar_to_cloud_node" name="sonar_to_cloud" output="screen" />

    <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
        <rosparam file="$(find obstacle_avoidance)/config/costmap_common_params.yaml" command="load" ns="global_costmap" />
        <rosparam file="$(find obstacle_avoidance)/config/costmap_common_params.yaml" command="load" ns="local_costmap" />

        <rosparam file="$(find obstacle_avoidance)/config/local_costmap_params.yaml" command="load" />
        <rosparam file="$(find obstacle_avoidance)/config/global_costmap_params.yaml" command="load" />

        <rosparam file="$(find obstacle_avoidance)/config/dwa_local_planner_params.yaml" command="load" />
        
        <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS" />
        
        <remap from="cmd_vel" to="/$(arg namespace)/cmd_vel_move_base"/>
    </node>
    
    <node pkg="obstacle_avoidance" type="depth_hold_shim.py" name="depth_hold_shim" output="screen" />
    <node pkg="obstacle_avoidance" type="send_navigation_goal.py" name="goal_sender" output="screen" />

    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find baseline)/rviz/navigation.rviz" />


    <include file="$(find rexrov2_gazebo)/launch/record.launch">
      <arg name="record" value="$(arg record)"/>
      <arg name="bag_filename" value="$(arg bag_filename)"/>
    </include>
</launch>

<!-- 
catkin_make

source devel/setup.bash

roslaunch obstacle_avoidance start.launch
 -->